scipy/stats/_distn_infrastructure.py

```diff
@@ -1257,16 +1262,18 @@ def moment(self, n, *args, **kwds):
 
         # Convert to transformed  X = L + S*Y
         # E[X^n] = E[(L+S*Y)^n] = L^n sum(comb(n, k)*(S/L)^k E[Y^k], k=0...n)
-        if loc == 0:
-            return scale**n * val
-        else:
-            result = 0
-            fac = float(scale) / float(loc)
-            for k in range(n):
-                valk = _moment_from_stats(k, mu, mu2, g1, g2, self._munp, args)
-                result += comb(n, k, exact=True)*(fac**k) * valk
-            result += fac**n * val
-            return result * loc**n
```

Comment:
I don't think there's anything wrong here, but as the one who might merge this, here's my preference:

Since this is a really fundamental function, I'd like to see how little we can change these lines. It would be great if you could even keep the original variable name `result`. If you implement @pvanmulbregt's suggestion to keep only the elements that meet both `cond`itions, I think you can keep these lines almost exactly the same as they are. They're pretty close right now, but please even avoid the refactoring (e.g. keep the `+=` as they are, etc.)

Reply:
> creating a result array of zeros only large enough to hold the results for the elements that meet both conditions - the ones that need to go through the iterative procedure

done.

> minimal changes to the iterative procedure

done.

> reassembling the output - putting in nans where cond0 wasn't met and putting scale**n * val where cond1 wasn't met

done.

I think that the current implementation should work fine for all array inputs too.